core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,266 @@
+    val queue = mutable.Queue.empty[Int]
+    queue += pid
+    while( !queue.isEmpty ) {
+      val p = queue.dequeue()
+      val c = getChildPids(p)
+      if(!c.isEmpty) {
+        queue ++= c
+        ptree += (p -> c.toSet)
+      }
+      else {
+        ptree += (p -> Set[Int]())
+      }
+    }
+  }
+
+  private def getChildPids(pid: Int): ArrayBuffer[Int] = {
+    try {
+      val cmd = Array("pgrep", "-P", pid.toString)
+      val input = Runtime.getRuntime.exec(cmd).getInputStream
+      val childPidsInByte = mutable.ArrayBuffer.empty[Byte]
+      var d = input.read()
+      while (d != -1) {
+        childPidsInByte.append(d.asInstanceOf[Byte])
+        d = input.read()
+      }
+      input.close()
+      val childPids = new String(childPidsInByte.toArray, "UTF-8").split("\n")
+      val childPidsInInt = mutable.ArrayBuffer.empty[Int]
+      for (p <- childPids) {
+        if (p != "") {
+          childPidsInInt += Integer.parseInt(p)
+        }
+      }
+      childPidsInInt
+    } catch {
+      case e: IOException => logDebug("IO Exception when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped", e)
+        isAvailable = false
+        return mutable.ArrayBuffer.empty[Int]
+    }
+  }
+
+  def getProcessInfo(pid: Int): Unit = {
+    /*
+   * Hadoop ProcfsBasedProcessTree class used regex and pattern matching to retrive the memory
+   * info. I tried that but found it not correct during tests, so I used normal string analysis
+   * instead. The computation of RSS and Vmem are based on proc(5):
+   * http://man7.org/linux/man-pages/man5/proc.5.html
+   */
+    try {
+      val pidDir = new File(procfsDir, pid.toString)
+      val fReader = new InputStreamReader(
+        new FileInputStream(
+          new File(pidDir, procfsStatFile)), Charset.forName("UTF-8"))
+      val in = new BufferedReader(fReader)
+      val procInfo = in.readLine
+      in.close
+      fReader.close
+      val procInfoSplit = procInfo.split(" ")
+      if ( procInfoSplit != null ) {
+        val vmem = procInfoSplit(22).toLong
+        val rssPages = procInfoSplit(23).toLong
+        if (procInfoSplit(1).toLowerCase(Locale.US).contains("java")) {
+          latestJVMVmemTotal += vmem
+          latestJVMRSSTotal += rssPages
+        }
+        else if (procInfoSplit(1).toLowerCase(Locale.US).contains("python")) {
+          latestPythonVmemTotal += vmem
+          latestPythonRSSTotal += rssPages
+        }
+        else {
+        latestOtherVmemTotal += vmem
+        latestOtherRSSTotal += rssPages }
+      }
+    } catch {
+      case f: FileNotFoundException => log.debug("There was a problem with reading" +
+        " the stat file of the process", f)
+    }
+  }
+
+  def updateAllMetrics(): Unit = {
+    allMetrics = computeAllMetrics
+  }
+
+  private def computeAllMetrics(): ProcfsBasedSystemsMetrics = {
+    if (!isAvailable) {
+      return ProcfsBasedSystemsMetrics(-1, -1, -1, -1, -1, -1)
```

Comment:
Setting a 0 for all of them should be fine here and will also remove the need to check if a proc based file system is available.